### PR TITLE
MRG: Minor fixes to testing framework

### LINKED ISSUE #4518: Can't run unittests, since '-a' is not recognized as a n option by my version of pytest
When I try to run unittests, I get the following error:

```
$ make test
python setup.py build_ext -i
running build_ext
rm -f .coverage
py.test -a 'not ultra_slow_test' mne
usage: py.test [options] [file_or_dir] [file_or_dir] [...]
py.test: error: unrecognized arguments: -a
  inifile: None
  rootdir: ~/Downloads/mne-python
Makefile:41: recipe for target 'test' failed
make: *** [test] Error 2
```
The version of py.test (which is seemingly the same as pytest !?) I have installed is:

```
$ py.test --version
This is pytest version 3.0.7, imported from /usr/local/lib/python2.7/dist-packages/pytest.pyc
setuptools registered plugins:
  pytest-xdist-1.20.0 at /usr/local/lib/python2.7/dist-packages/xdist/looponfail.py
  pytest-xdist-1.20.0 at /usr/local/lib/python2.7/dist-packages/xdist/plugin.py
  pytest-forked-0.2 at /usr/local/lib/python2.7/dist-packages/pytest_forked/__init__.pyc
  pytest-cov-2.5.1 at /usr/local/lib/python2.7/dist-packages/pytest_cov/plugin.py

```

When I run it with --help, I get this:

```
$ py.test --help
usage: py.test [options] [file_or_dir] [file_or_dir] [...]

positional arguments:
  file_or_dir

general:
  -k EXPRESSION         only run tests which match the given substring
                        expression. An expression is a python evaluatable
                        expression where all names are substring-matched
                        against test names and their parent classes. Example:
                        -k 'test_method or test_other' matches all test
                        functions and classes whose name contains
                        'test_method' or 'test_other'. Additionally keywords
                        are matched to classes and functions containing extra
                        names in their 'extra_keyword_matches' set, as well as
                        functions which have names assigned directly to them.
  -m MARKEXPR           only run tests matching given mark expression.
                        example: -m 'mark1 and not mark2'.
  --markers             show markers (builtin, plugin and per-project ones).
  -x, --exitfirst       exit instantly on first error or failed test.
  --maxfail=num         exit after first num failures or errors.
  --strict              run pytest in strict mode, warnings become errors.
  -c file               load configuration from `file` instead of trying to
                        locate one of the implicit configuration files.
  --continue-on-collection-errors
                        Force test execution even if collection errors occur.
  --fixtures, --funcargs
                        show available fixtures, sorted by plugin appearance
  --fixtures-per-test   show fixtures per test
  --import-mode={prepend,append}
                        prepend/append to sys.path when importing test
                        modules, default is to prepend.
  --pdb                 start the interactive Python debugger on errors.
  --pdbcls=modulename:classname
                        start a custom interactive Python debugger on errors.
                        For example:
                        --pdbcls=IPython.terminal.debugger:TerminalPdb
  --capture=method      per-test capturing method: one of fd|sys|no.
  -s                    shortcut for --capture=no.
  --runxfail            run tests even if they are marked xfail
  --lf, --last-failed   rerun only the tests that failed at the last run (or
                        all if none failed)
  --ff, --failed-first  run all tests but run the last failures first. This
                        may re-order tests and thus lead to repeated fixture
                        setup/teardown
  --cache-show          show cache contents, don't perform collection or tests
  --cache-clear         remove all cache contents at start of test run.

reporting:
  -v, --verbose         increase verbosity.
  -q, --quiet           decrease verbosity.
  -r chars              show extra test summary info as specified by chars
                        (f)ailed, (E)error, (s)skipped, (x)failed, (X)passed,
                        (p)passed, (P)passed with output, (a)all except pP.
                        The pytest warnings are displayed at all times except
                        when --disable-pytest-warnings is set
  --disable-pytest-warnings
                        disable warnings summary, overrides -r w flag
  -l, --showlocals      show locals in tracebacks (disabled by default).
  --tb=style            traceback print mode (auto/long/short/line/native/no).
  --full-trace          don't cut any tracebacks (default is to cut).
  --color=color         color terminal output (yes/no/auto).
  --durations=N         show N slowest setup/test durations (N=0 for all).
  --pastebin=mode       send failed|all info to bpaste.net pastebin service.
  --junit-xml=path      create junit-xml style report file at given path.
  --junit-prefix=str    prepend prefix to classnames in junit-xml output
  --result-log=path     DEPRECATED path for machine-readable result log.

collection:
  --collect-only        only collect tests, don't execute them.
  --pyargs              try to interpret all arguments as python packages.
  --ignore=path         ignore path during collection (multi-allowed).
  --confcutdir=dir      only load conftest.py's relative to specified dir.
  --noconftest          Don't load any conftest.py files.
  --keep-duplicates     Keep duplicate tests.
  --doctest-modules     run doctests in all .py modules
  --doctest-report={none,cdiff,ndiff,udiff,only_first_failure}
                        choose another output format for diffs on doctest
                        failure
  --doctest-glob=pat    doctests file matching pattern, default: test*.txt
  --doctest-ignore-import-errors
                        ignore doctest ImportErrors

test session debugging and configuration:
  --basetemp=dir        base temporary directory for this test run.
  --version             display pytest lib version and import information.
  -h, --help            show help message and configuration info
  -p name               early-load given plugin (multi-allowed). To avoid
                        loading of plugins, use the `no:` prefix, e.g.
                        `no:doctest`.
  --trace-config        trace considerations of conftest.py files.
  --debug               store internal tracing debug information in
                        'pytestdebug.log'.
  -o [OVERRIDE_INI [OVERRIDE_INI ...]], --override-ini=[OVERRIDE_INI [OVERRIDE_INI ...]]
                        override config option with option=value style, e.g.
                        `-o xfail_strict=True`.
  --assert=MODE         Control assertion debugging tools. 'plain' performs no
                        assertion debugging. 'rewrite' (the default) rewrites
                        assert statements in test modules on import to provide
                        assert expression information.
  --setup-only          only setup fixtures, do not execute tests.
  --setup-show          show setup of fixtures while executing tests.
  --setup-plan          show what fixtures and tests would be executed but
                        don't execute anything.

distributed and subprocess testing:
  -f, --looponfail      run tests in subprocess, wait for modified files and
                        re-run failing test set until all pass.
  -n numprocesses, --numprocesses=numprocesses
                        shortcut for '--dist=load --tx=NUM*popen', you can use
                        'auto' here for auto detection CPUs number on host
                        system
  --max-slave-restart=MAX_SLAVE_RESTART
                        maximum number of slaves that can be restarted when
                        crashed (set to zero to disable this feature)
  --dist=distmode       set mode for distributing tests to exec environments.
                        each: send each test to all available environments.
                        load: load balance by sending any pending test to any
                        available environment. loadscope: load balance by
                        sending pending groups of tests in the same scope to
                        any available environment. (default) no: run tests
                        inprocess, don't distribute.
  --tx=xspec            add a test execution environment. some examples: --tx
                        popen//python=python2.5 --tx socket=192.168.1.102:8888
                        --tx ssh=user@codespeak.net//chdir=testcache
  -d                    load-balance tests. shortcut for '--dist=load'
  --rsyncdir=DIR        add directory for rsyncing to remote tx nodes.
  --rsyncignore=GLOB    add expression for ignores when rsyncing to remote tx
                        nodes.
  --boxed               backward compatibility alias for pytest-forked
                        --forked

forked subprocess test execution:
  --forked              box each test run in a separate process (unix)

coverage reporting with distributed testing support:
  --cov=[path]          measure coverage for filesystem path (multi-allowed)
  --cov-report=type     type of report to generate: term, term-missing,
                        annotate, html, xml (multi-allowed). term, term-
                        missing may be followed by ":skip-covered". annotate,
                        html and xml may be followed by ":DEST" where DEST
                        specifies the output location. Use --cov-report= to
                        not generate any output.
  --cov-config=path     config file for coverage, default: .coveragerc
  --no-cov-on-fail      do not report coverage if test run fails, default:
                        False
  --no-cov              Disable coverage report completely (useful for
                        debuggers) default: False
  --cov-fail-under=MIN  Fail if the total coverage is less than MIN.
  --cov-append          do not delete coverage but append to current, default:
                        False
  --cov-branch          Enable branch coverage.


[pytest] ini-options in the first pytest.ini|tox.ini|setup.cfg file found:

  markers (linelist)       markers for test functions
  norecursedirs (args)     directory patterns to avoid for recursion
  testpaths (args)         directories to search for tests when no files or directories are given in the command line.
  usefixtures (args)       list of default fixtures to be used with this project
  python_files (args)      glob-style file patterns for Python test module discovery
  python_classes (args)    prefixes or glob names for Python test class discovery
  python_functions (args)  prefixes or glob names for Python test function and method discovery
  xfail_strict (bool)      default for the strict parameter of xfail markers when not given explicitly (default: False)
  doctest_optionflags (args) option flags for doctests
  addopts (args)           extra command line options
  minversion (string)      minimally required pytest version
  rsyncdirs (pathlist)     list of (relative) paths to be rsynced for remote distributed testing.
  rsyncignore (pathlist)   list of (relative) glob-style paths to be ignored for rsyncing.
  looponfailroots (pathlist) directories to check for changes

environment variables:
  PYTEST_ADDOPTS           extra command line options
  PYTEST_PLUGINS           comma-separated plugins to load during startup
  PYTEST_DEBUG             set to enable debug tracing of pytest's internals


to see available markers type: pytest --markers
to see available fixtures type: pytest --fixtures
(shown according to specified file_or_dir or current dir if not specified)

```
Since this doesn't seem to allow for the -a option from the Makefile command, I'm wondering was the '-a' a typo, or do I need to install some other version of py.test/pytest?

### PR DESCRIPTION
**Pull request by larsoner**
This PR:

1. Gets rid of some `nose` cruft
2. Moves to standard `pytest` slow-test markers
3. Fixes our decorators for `sklearn` that had incorrect versions
4. Makes AppVeyor a little bit nicer

Closes #4518.

### CODE REVIEW

**Comment by larsoner** [comment]
@agramfort CIs look like they're headed toward happiness

File: .travis.yml [hunk 1]
```diff
@@ -13,78 +13,78 @@ matrix:
     include:
         # Full (Linux, 2.7) first half
         - os: linux
-          env: CONDA_DEPENDENCIES="numpy scipy matplotlib pandas scikit-learn h5py pillow statsmodels mayavi pytest"
-               PIP_DEPENDENCIES="git+git://github.com/nipy/PySurfer.git nitime faulthandler joblib nibabel codecov nose-timer nose-faulthandler pytest-sugar pytest-cov pytest-attrib"
+          env: CONDA_DEPENDENCIES="numpy scipy matplotlib pandas scikit-learn h5py pillow statsmodels mayavi nose pytest pytest-cov"
```
Comment (agramfort) [inline .travis.yml]: remove nose?

**Comment by agramfort** [comment]
besides my concern of keeping some nose in travis LGTM

thx heaps @larsoner 

@massich wanna have a look?

**Comment by larsoner** [comment]
No we still do a lot of nose assert imports. It will take a separate PR
with thousands of changed lines to fix that

**Comment by agramfort** [comment]
how about having our assert_* functions in fixes.py so we don't need to change every single line for now?

**Comment by larsoner** [comment]
Only hundreds of lines then :)

Probably better for a follow up PR though

**Comment by agramfort** [comment]
yes ok.

**Comment by massich** [comment]
LGTM

**Comment by massich** [comment]
> how about having our assert_* functions in fixes.py so we don't need to change every single line for now?

I would do that as intermediate step if we really need to remove the nose dependency. But as end goal I would change all `assert_*` to **pytest** like asserts.

**Comment by agramfort** [comment]
@larsoner you need to rebase

**Comment by larsoner** [comment]
@agramfort done

**Comment by agramfort** [comment]
is the Makefile and the make test target already switched to pytest ?

if so +1 for MRG

**Comment by larsoner** [comment]
Yep, @massich did that in a previous PR and I tweaked the values here based on the attribute/mark switch

Just pushed a commit to fix a rebase error, will merge once the CIs come back happy

Then we can figure out #4523 :) Maybe the first thing to do is just put the existing Travis 2.7 config in `environment.yml`. That way we'll have something that works on CIs, even if it doesn't work well for users. Then we can iterate on just modifying that file (and/or adding a separate 3.5/6 one) without needing to mess very much with Travis or AppVeyor.

**Comment by larsoner** [comment]
Going to have to rebase again. @agramfort okay to merge once CIs are happy?

**Comment by agramfort** [comment]
yes ok to merge when Cis are green

**Pull request merged**